stan/math/prim/meta/is_eigen.hpp

```diff
@@ -2,42 +2,41 @@
 #define STAN_MATH_PRIM_META_IS_EIGEN_HPP
 
 #include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/meta/is_eigen_matrix.hpp>
 #include <stan/math/prim/meta/disjunction.hpp>
 #include <stan/math/prim/meta/scalar_type.hpp>
 #include <stan/math/prim/meta/value_type.hpp>
 #include <type_traits>
 
 namespace stan {
 
-/** \ingroup type_trait
- * Base implementation to check whether a type is derived from EigenBase
+/** \addtogroup type_trait
+ *  @{
  */
-template <typename T, typename = void>
-struct is_eigen : std::false_type {};
 
-namespace internal {
 /**
- * Underlying implementation to check if a type is derived from EigenBase
- */
-template <typename T>
-struct is_eigen_base {
-  static std::false_type f(const void *);
-  template <typename Derived>
-  static std::true_type f(const Eigen::EigenBase<Derived> *);
-  enum { value = decltype(f(std::declval<T *>()))::value };
-};
+ * Check if type derives from `EigenBase`
+ * @tparam T Type to check if it is derived from `EigenBase`
+ * @tparam Enable used for SFINAE deduction.
+ **/
+template <typename T, typename Enable = void>
+struct is_eigen : std::false_type {};
 
-}  // namespace internal
+template <typename T>
+struct is_eigen<T, std::enable_if_t<std::is_base_of<
+                       Eigen::EigenBase<typename std::decay_t<T>::PlainObject>,
+                       typename std::decay_t<T>::PlainObject>::value>>
+    : std::true_type {};
 
-/**
- * Checks whether type T is derived from EigenBase. If true this will have a
- * static member function named value with a type of true, else value is false.
- */
 template <typename T>
-struct is_eigen<
-    T, std::enable_if_t<internal::is_eigen_base<std::decay_t<T>>::value>>
+struct is_eigen<T, std::enable_if_t<std::is_base_of<
+                       Eigen::EigenBase<typename std::decay_t<T>::MatrixType>,
+                       typename std::decay_t<T>::MatrixType>::value>>
     : std::true_type {};
 
+template <typename T>
+struct is_eigen<Eigen::EigenBase<T>, void> : std::true_type {};
+
```

Comment:
The new type traits check that the type inherits from a base class, but `EigenBase` is the base class everything inherits from and it does not inherit from itself. So this particular type trait needs an actual specialization for the base class.

Reply:
Here's an example

https://godbolt.org/z/O9O4DX

Reply:
Would assignability to `EigenBase` be a way to implemen this?

Reply:
`EigenBase` is not really something that is assigned to. The base class for Eigen essentially just defines their lowest level API that all their other classes make available.

I thought about defining these by assignability, but it starts getting a bit weird. Like [here](https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/EigenBase.h#L134) where they have assignment for `EigenBase` derived expressions to `DenseBase` objects. The way they use CRTP is to define layers of APIs so for our purposes what we want is to check that an object inherits from a particular class so that we can use that class's API.

https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/EigenBase.h

Reply:
> The new type traits check that the type inherits from a base class

Why would we nee to change the definition of `is_eigen` in this way? That example only shows a bug in a halfway changed implementation, not in existing one.

> Would assignability to EigenBase be a way to implemen this?

Yes, but `EigenBase` is a template. To check assignability we would still need to decide on template arguments. It would probably allow to remove one case from the implementation suggested in this PR. But I still did not see any arguments why we need to change it at all.

> I thought about defining these

The issue here is change of `is_eigen`. The suggested implementation is both longer, more convoluted and has some weird issues as seen in the example on https://godbolt.org/z/7xxK9n. Other type traits can be discussed separately.

Reply:
Actually ... maybe we could implement all or most type traits by checking for pointer conversions. Here is an example for `is_eigen_matrix_base` : https://godbolt.org/z/or6NPL

Reply:
Huh! That's a really neat approach! Let me fiddle around with this. I like it's end result but it's a bit confusing (though tbh so is the current thing I'm proposing)